(TK-172) Add endpoint to query status for a single service

rlinehan opened the pull request:
Add the ability to query for a single service's status at
/status/v1/services/<service name>. If the service has multiple status
callback functions registered, the endpoint returns the result of calling the
latest (same as for the rollup endpoint). If the service is nonexistent/hasn't
registered itself with the status service, the endpoint returns a 404.

rlinehan committed:
(TK-172) Add endpoint to query status for a single service

Add the ability to query for a single service's status at
/status/v1/services/<service name>. If the service has multiple status
callback functions registered, the endpoint returns the result of calling the
latest (same as for the rollup endpoint). If the service is nonexistent/hasn't
registered itself with the status service, the endpoint returns a 404.

rlinehan commented on test/puppetlabs/trapperkeeper/services/status/status_service_test.clj:
```diff
@@ -45,3 +45,27 @@
+        (is (= 200 (:status req)))
+        (is (= {"service-version" "1.1.0"
+                "service-status-version" 2
+                "status" "foo status 2"}
```
The only thing I wasn't sure about here was whether it would be useful to also return the service's name in the body?

jeffmccune replied:
Name might be useful to allow consumers to interact only with the data returned, not their own context or state.

MSLilah replied:
I'd lean towards including it. I think the user can definitely get by without it since they'll need it anyway to query a specific service's status, but I think it might be a little more convenient to have it in the body.

MSLilah commented:
:+1:

rlinehan committed:
(TK-172) Add service name to response when querying single service

rlinehan committed:
(maint) Remove outer level of nesting for rollup endpoint response

Previously, the rollup endpoint would return output in the form

    {"services" {"foo" {...}
                 "bar" {...}}}

This nesting under "services" is unnecessary because there is nothing else in
this map, so this commit removes it.

rlinehan commented:
Added a commit adding the service name to the response body for the single service endpoint.

Also added a commit removing the outer map which only contained one key (`services`) from the rollup endpoint.

KevinCorcoran commented on src/puppetlabs/trapperkeeper/services/status/status_core.clj:
```diff
@@ -68,7 +68,16 @@
-           :body {"services" statuses}})))))
+           :body statuses}))
+       (compojure/GET "/services/:service-name" [service-name]
+         (if-let [status-info (get (deref status-fns-atom) service-name)]
```
I think a better name would be `service-info`, the corresponding schema is `ServiceInfo`.

KevinCorcoran commented on src/puppetlabs/trapperkeeper/services/status/status_core.clj:
```diff
@@ -68,7 +68,16 @@
+             {:status 200
+              :body (assoc status :service-name service-name)})
+           {:status 404
+            :body {:error {:type ::service-not-found
```
I'm not sure if I like namespaces ending up in the response.  Perhaps just `:service-not-found` here instead?

rlinehan replied:
I agree, this is just what we've done in other places (e.g. file sync)?

KevinCorcoran replied:
ORLY?  Well, I'd be happy to change it in file sync, and I wouldn't consider anything in that repo to be canon at this point.

rlinehan replied:
https://github.com/rlinehan/pe-file-sync/commit/fff3b910900f14452a6faba23042f73e3b859849#commitcomment-10500725 and https://github.com/rlinehan/pe-file-sync/commit/dfc313bf473b57e1dd3000d5c5ebf44ed0bb70e9, although in that case I was kind of making up a partial namespace. I think sometimes having the response indicate what trapperkeeper service an error came from could maybe be useful in the case of http responses (e.g. `:status-service/service-not-found`), but I don't like having the full namespace in the response - I'd prefer just `:service-not-found` rather than the full namespace.

KevinCorcoran replied:
Hmm, yeah, I can see something like that being useful in general, but ...

In this specific case, a user doesn't care about the status service at all - hopefully, they don't even know it exists.  They'll be querying for the status of an individual service they care about, e.g. Puppet Server, so anything about `status-service` in the response is probably just noise (at least in this use-case).

I feel like we're treading on :bike: territory here ... I vote for just `:service-not-found` but since we don't have a hard requirement one way or the other at this point, I'll be fine with whatever you want to go with.

KevinCorcoran commented:
:+1: Looks fine overall.

rlinehan committed:
(TK-172) Cleanup: rename parameter, don't namespace error type

Rename `status-info` parameter to `service-info` to match the schema name.

Remove namespacing of error type - we don't really want to expose the full
namespace in the http response.

rlinehan commented:
@KevinCorcoran added a commit addressing your comments.
